### DISCUSSION TITLE
Using PreserveReferences() throws an error for mapping of source that is Value type

### DISCUSSION THREAD

**Discussion by BogdanGirnyk** [LuckyPennySoftware/AutoMapper#3871] [NA]
Hi,
I faced an issue when creating a map from Enum to some class using `.PreserveReferences()` .
```
var config = new MapperConfiguration(cfg => 
{
    cfg.CreateMap<SomeEnum, DestinationClass>()
        .ForMember(dest => dest.Id, opt => opt.MapFrom(src => src))
        .PreserveReferences();
});
```
The expression  throws an exception:

> System.ArgumentException: 'Expression of type 'TestConsole.Models.SomeEnum' cannot be used for parameter of type 'System.Object' of method 'Void CacheDestination(System.Object, System.Type, System.Object)' (Parameter 'arg0')'

Looks like behavior is the same for any ValueType destination.
Is it expected or it is a bug?

**Comment by lbargaoanu** [comment]
It doesn't make sense for value types, they are values, not references.

**Reply by BogdanGirnyk** [reply]
But I want the same enum values to be mapped to same destination references.
Here is my situation:
Source model:
```
public class Source
{
    public int Id { get; set; } 

    public SomeEnum Enum { get; set; }
}

public enum SomeEnum
{
    first,
    second,
    third
}
```
Destination model:
```
public class Destination
{
    public int Id { get; set; }

    public InnerClass DestinationInner { get; set; }
}

public class InnerClass
{
    public SomeEnum Id { get; set; }
}
```
Mapping:
```
cfg.CreateMap<SomeEnum, InnerClass>()
    .ForMember(dest => dest.Id, opt => opt.MapFrom(src => src));
//Fails here
//.PreserveReferences();
cfg.CreateMap<Source, Example.Destination>()
    .ForMember(dest => dest.Id, opt => opt.MapFrom(src => src.Id))
    .ForMember(dest => dest.DestinationInner, opt => opt.MapFrom(src => src.Enum))
    .PreserveReferences();
```
Act:
```
List<Source> sources = new List<Source>
{
    new Source { Id = 1, Enum = SomeEnum.first},
    new Source { Id = 2, Enum = SomeEnum.first},
};

var result = mapper.Map<List<Destination>>(sources);

var inner1 = result[0].DestinationInner;
var inner2 = result[1].DestinationInner;

// it should be true
var oneEqTwoRef = ReferenceEquals(inner1, inner2);
```
PreserveReferences() approach works perfectly for strings and other reference types to achieve reference equality, but it fails for Value types.
Maybe you could advice another way?
Thanks in advance.

**Reply by jbogard** [reply]
I think you have a fundamental misunderstanding of value types versus reference types. Value types are ALWAYS copied:
```
int i, j;
i = 5;
j = i;
var isEqual = ReferenceEquals(i, j); // this is ALWAYS false for value types
```

**Reply by BogdanGirnyk** [reply]
I think I got it now.
The result references will be equal only if the source references are equal.
Thank you!
